course/latex/latex.py

```diff
@@ -0,0 +1,176 @@
+        has_ele = []
+
+        for ele, pattern in ele_re_tuple:
+            required_ele_list.append(ele)
+            iter = re.finditer(pattern, latex)
+
+            matched_indice = [m.start(0) for m in iter]
+            matched_len = len(matched_indice)
+            if matched_len == 0:
+                if not test:
+                    raise TexDocMissingElementError(
+                        _("No %s found in latex source") % ele)
+                else:
+                    has_ele.append(False)
+            elif matched_len > 1:
+                raise TexDocParseError(
+                    _("More than one %s found in latex source") % ele)
+            else:
+                if test:
+                    has_ele.append(True)
+                ele_position_list.append(matched_indice[0])
+
+        if test:
+            [self.has_preamble, self.has_begindoc, self.has_enddoc] = has_ele
+
+        if not ele_position_list == sorted(ele_position_list):
+            raise TexDocWrongElementOrderError(
+                _("The occurance of %s are not in proper order")
+                % ",".join(required_ele_list))
+
+        if not test:
+            [preamble, document] = latex.split((r"\begin{document}"))
+            document = document.split((r"\end{document}"))[0]
+            self.preamble = strip_spaces(preamble)
+            self.document = strip_spaces(document, allow_single_empty_line=True)
+            assert self.preamble is not None
+            assert self.document is not None
+
+    def as_latex(self):
+        """
+        Assemble LaTeX Document
+        """
+        latex = ""
+        if self.empty_pagestyle:
+            if not self.is_empty_pagestyle_already():
+                self.preamble += "\n\\pagestyle{empty}\n"
+
+        latex += self.preamble
+        latex += "\\begin{document}\n"
+        latex += self.document
+        latex += "\\end{document}\n"
+
+        return latex
+
+    def __str__(self):
+        return self.document()
+
+    def __unicode__(self):
+        return self.document()
+
+    def __init__(self, text=None, preamble="", preamble_extra="",
+                 empty_pagestyle=False):
+        """
+        Parse LaTeX document
+        :param text: string. Full latex document, or body only if
+        preamble or preamble_extra are given.
+        :param preamble: string. If full document is provided in
+        text, this value will be neglected.
+        :param preamble_extra: string. Append to existing preamle.
+        :param empty_pagestyle: bool. If True, the pagestyle will
+        be set as "empty". We are not using
+        \documentclass{standalone}.
+        """
+        if text:
+            text = strip_comments(text)
+            try:
+                self.parse(text)
+            except TexDocMissingElementError:
+                self.parse(text, test=True)
+                if self.has_preamble:
+                    # begin_document or end_document is missing
```

Comment:
Because ``TexDocMissingElementError`` will be raise if only one of ``\documentclass``, ``\begin{document}`` and ``\end{document}`` is missing.